### ISSUE TITLE
<dom-if> has a property propagation bug

### ISSUE THREAD

**Issue by MajorBreakfast** [Polymer/polymer#5094]
### Description
The idea behind the current implementation of `<dom-if>` is to not propagate property changes from the host to the template instance while the condition is `false`. Then, when the condition becomes `true` again, all property changes get applied at once. The problem is that subtle changes like array mutations get lost along the way.

#### Live Demo
https://codepen.io/MajorBreakfast/pen/OQXGrm

https://codepen.io/MajorBreakfast/pen/MQjWxo (Edit: Demo without `<dom-repeat>`)

#### Steps to Reproduce
- Create a `<dom-repeat>` inside a `<dom-if>`
- Set `if` condition to `true`, so that `<dom-if>`'s template gets instantiated
- Wait a bit for `<dom-repeat>` to render its items
- Set `if` condition to `false`
- Mutate the array that `<dom-repeat>` uses using `PolymerElement#push()`
- Set `if` condition to `true` again

#### Expected Results
`<dom-repeat>` should display the array with the mutation that just happened applied

#### Actual Results
`<dom-repeat>` does not display the mutated array. It does not update because it is not told properly about the array mutation. It would only catch the change if it had mutable data turned on. So, there is different behavior depending on whether a change happens while the condition is `true` or `false`. This difference is introduced by `<dom-if>`'s optimization to delay the propagation of properties until its content becomes visible again.

### Browsers Affected
- [x] Chrome
- [x] Firefox
- [x] Edge
- [x] Safari 9
- [x] Safari 8
- [x] IE 11

### Versions
- Polymer: v2.5
- webcomponents: v1.1.0

@sorvell 
@kevinpschaaf

**Comment by MajorBreakfast** [comment]
One way to make this work is to modify how the `MutableData` mixin works (`<dom-if>` sets the mutableData option when it calls `templatize()`). If the change was somehow marked that it originates from a source that has mutable data activated, then all resulting modifications could use the mutable data comparison strategy instead of strict equality. Using mutable data would fix this. The `<dom-repeat>` that's nested inside `<dom-if>` just doesn't know that it needs to use it temporarily to catch the change. This is of course not a problem with `<dom-repeat>`, instead any nested element has this problem.

**Comment by MajorBreakfast** [comment]
@TimvdLippe You can remove the `dom-repeat` tag. `<dom-repeat>` is not at fault.

Here's how this bug looks like without `<dom-repeat>`:
https://codepen.io/MajorBreakfast/pen/MQjWxo

**Comment by MajorBreakfast** [comment]
I did some thinking about this and here's what I come up with:
Any delay to the propagation of deep changes doesn't make much sense because the underlying object gets mutated immediately either way. Not announcing the change when it happens just leads to bugs. 

Here's an example:
```JS
prop = { a: { b: 'v1' } }
```

```HTML
<x-element my-value="{{prop.a}}"></x-element>
```
- Delaying the propagation of changes to `prop` or `prop.a` would be fine. Giving `<x-element>` the new data a bit later is fine
- In contrast, delaying the propagation of a change to `prop.a.b` isn't fine because the synced object (object in `a`) is mutated whether we tell `<x-element>` about it or not

Also, I think that this whole logic shouldn't live inside `<dom-if>`. Instead it could be a feature of the template instance created by `templatize()`. Any here's why:
- Only the template knows which concrete paths are synced (`'prop.a'` in the example above)
- The synchronization in the other direction (template instance -> host) should be stopped as well. But ,again, only for non-deep changes. Deep changes should always be propagated because the objects are mutated.

Or... maybe Polymer should drop this "feature" entirely. IMO it just introduces unreliability into a reliable system.

TL;DR I think the delayed synchronization should be removed from `<dom-if>`. At least in its current buggy form :)

**Comment by TimvdLippe** [comment]
@MajorBreakfast could you confirm/deny that this is a duplicate of #4818?

**Comment by MajorBreakfast** [comment]
@TimvdLippe Indeed. It is a duplicate.

They only way to monkey patch this broken behavior is to add the `MutableData` mixin to every element inside the `<dom-if>` (or nested inside nested elements, etc.). That works of course only for code that is under your control. Here's a "fixed" version of the demo of the other issue: http://jsbin.com/zuwohoqeyo/edit?html,console,output

**Comment by MajorBreakfast** [comment]
Reliability shouldn't be opt-in.

But, it could be opt-out: The optimization could be made optional. Set the `no-sync-while-false` attribute and `<dom-if>` delays any property changes until the condition is `true` again.

**Comment by TimvdLippe** [comment]
All right thanks for confirming @MajorBreakfast ! I am closing this one as duplicate of #4818. Feel free to add any of your suggestions to #4839, which is fixes this underlying issue.
